src/main/java/org/elasticsearch/common/geo/builders/BasePolygonBuilder.java

```diff
@@ -112,6 +116,18 @@ public ShapeBuilder close() {
     }
 
     /**
+     * Validates only 1 vertex is tangential (shared) between the interior and exterior of a polygon
+     */
+    protected void validateHole(BaseLineStringBuilder shell, BaseLineStringBuilder hole) {
+        ArrayList exterior = Lists.newArrayList(Sets.newHashSet(shell.points));
```

Comment:
Actually, it looks like retainAll always uses contains (I was thinking it could be smarter with a sorted list, but of course it has no way to know it is sorted!).  So I would copy, sort, then just iterate both in step?